fedora-ostree-pruner/fedora-ostree-pruner

```diff
@@ -0,0 +1,278 @@
+    runcmd(cmd)
+
+
+@catch_exceptions_and_continue
+def prune_compose_repo(test=False):
+    # prune the compose repo
+    logger.info(f'Pruning the compose repo {OSTREECOMPOSEREPO}'
+                f' with policy {COMPOSE_REPO_POLICY}')
+    prunearg = pruning_policy_to_args(COMPOSE_REPO_POLICY)
+    cmd = ['ostree', 'prune', '--repo', OSTREECOMPOSEREPO,
+           '--refs-only', prunearg]
+    if test:
+        cmd.append('--no-prune')
+    runcmd(cmd)
+
+
+@catch_exceptions_and_continue
+def prune_prod_repo(test=False):
+    # Error out if any refs exist that aren't defined in the policy
+    cmd = ['ostree', 'refs', '--repo', OSTREEPRODREPO]
+    cp = runcmd(cmd, capture_output=True)
+    prod_refs = cp.stdout.decode('utf-8').splitlines()
+    for ref in prod_refs:
+        if ref not in PROD_REF_POLICIES:
+            msg = f'Ref {ref} in repo {OSTREEPRODREPO} but no policy defined'
+            logger.error(msg)
+            raise Exception(msg)
+
+    # Warn up front for refs that are in the policy but not in the repo
+    for ref, policy in PROD_REF_POLICIES.items():
+        if ref not in prod_refs:
+            if policy != 'delete':
+                logger.warning('Policy defined for a ref that is not in the'
+                               f' repo {OSTREEPRODREPO}: {ref}')
+            continue
+
+    # prune each branch in the policy with specified value
+    for ref,policy in PROD_REF_POLICIES.items():
+
+        # Skip if there is no ref in the repo for this policy
+        if ref not in prod_refs:
+            continue
+
+        if policy is None:
+            logger.info(f'Skipping ref {ref} in repo {OSTREEPRODREPO}.'
+                        'Policy is to keep all commits.')
+            continue
+
+        if policy == 'delete':
+            delete_ref_in_repo(ref, OSTREEPRODREPO, test)
+            continue
+
+        prunearg = pruning_policy_to_args(policy)
+
+        logger.info(f'Pruning the {ref} ref in repo {OSTREEPRODREPO} to {policy}')
+        # Use --commit-only so only the commit gets deleted and not related objects
+        # which take time to calculate reachability. We go through and first delete
+        # commits (fast) and then go back at the end and run a single generic delete
+        # prune of any unreachable objects in the repo (slow).
+        cmd = ['ostree', 'prune', '--repo', OSTREEPRODREPO,
+               '--commit-only', '--only-branch', ref, '--refs-only', prunearg]
+        if test:
+            cmd.append('--no-prune')
+        runcmd(cmd)
+
+    # Now run a generic prune once over the entire repo to delete any
+    # objects in the repo that aren't reachable. This should save some
+    # time since we only iterate over the entire repo once, versus for
+    # every ref/commit we deleted.
+    cmd = ['ostree', 'prune', '--repo', OSTREEPRODREPO, '--refs-only']
+    if test:
+        cmd.append('--no-prune')
+    runcmd(cmd)
+
+
+def catch_exceptions_and_continue(func):
```

Comment:
The only thing I'm not sure about is this. Right now we do:

```python
prune_compose_repo(...)
prune_prod_repo(...)
```

Is it the right thing to do to continue with pruning the prod repo if an error occurred in the compose repo?

Reply:
I think they're mostly independent, but probably depends on the error. i.e. if it's some sort of `I/O` error the problem will probably apply to both operations.